File: Medium Collection/BackTracking.java
1. Letter Combinations of a Phone Number

Given a string containing digits from 2-9 inclusive, return all possible letter combinations that the number could represent.

A mapping of digit to letters (just like on the telephone buttons) is given below. Note that 1 does not map to any letters.

Example:

Input: "23"
Output: ["ad", "ae", "af", "bd", "be", "bf", "cd", "ce", "cf"].

Note:

Although the above answer is in lexicographical order, your answer could be in any order you want.


import java.util.LinkedList;
import java.util.List;

class Solution {
    public static char[][] phone = new char[10][];

    Solution(){
        phone = new char[10][];
        phone[2] = new char[]{'a','b','c'};
        phone[3] = new char[]{'d','e','f'};
        phone[4] = new char[]{'g','h','i'};
        phone[5] = new char[]{'j','k','l'};
        phone[6] = new char[]{'m','n','o'};
        phone[7] = new char[]{'p','q','r','s'};
        phone[8] = new char[]{'t','u','v'};
        phone[9] = new char[]{'w','x','y','z'};
    }
   
    public List<String> letterCombinations(String digits) {
//        ArrayList<char[]> phone = new ArrayList<>(10);
//        phone.set(2,{"a","b","c"};

        char[] chars = digits.toCharArray();
        List<String> result = new LinkedList<>();
        if(digits==null || chars.length==0)  return result;
        result.add(digits);
        for(int i=0; i<digits.length(); i++){
            result = cycles(result,i);
        }
        return result;
    }
    
    //逐位替换数字为字母
    public List<String> cycle(String str, int index){
        char[] word = str.toCharArray();
        List<String> result =new LinkedList<>();
        char c = str.charAt(index);
        if(c>'9' || c<'0')  return result;
        char[] key = phone[c-'0'];
        for(int i=0; i<key.length; i++){
            word[index] = key[i];
            result.add(new String(word));
        }
        return result;
    }
    public List<String> cycles(List<String> strs, int index){
        List<String> result =new LinkedList<>();
        if(strs==null || strs.isEmpty())  return strs;
        for(String str : strs){
            List<String> one = cycle(str,index);
            result.addAll(one);
        }
        return result;
    }
}

2.Generate Parentheses

Given n pairs of parentheses, write a function to generate all combinations of well-formed parentheses.

For example, given n = 3, a solution set is:

[
  "((()))",
  "(()())",
  "(())()",
  "()(())",
  "()()()"
]

class Solution {

    public List<String> generateParenthesis(int n) {
        List<String> result = new LinkedList<>();
        if(n<=0)  return result;
        Set<String> set = new HashSet<>();
        set.add("()");
        for(int i=2; i<=n; i++){
            set = insert(set);
        }
        result.addAll(set);
        return result;
    }

    public Set<String> insert(Set<String> strs){
        Set<String> newset = new HashSet<>();
        for(String str:strs){
            for(int i=0; i<=str.length(); i++){
                StringBuilder sb = new StringBuilder(str);
                newset.add(new String(sb.insert(i,"()")));
            }
        }
       return newset;
    }

}

3. Permutations

Given a collection of distinct integers, return all possible permutations.

Example:

Input: [1,2,3]
Output:
[
  [1,2,3],
  [1,3,2],
  [2,1,3],
  [2,3,1],
  [3,1,2],
  [3,2,1]
]
3.1
public class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> result = new ArrayList<List<Integer>>();
        if (nums == null || nums.length == 0) return result;

        List<Integer> list = new ArrayList<Integer>();
        dfs(nums, list, result);

        return result;
    }

    private void dfs(int[] nums, List<Integer> list, List<List<Integer>> result) {
        if (list.size() == nums.length) {
            result.add(new ArrayList<Integer>(list));
            return;
        }

        for (int i = 0; i < nums.length; i++) {
            if (list.contains(nums[i])) continue;
            list.add(nums[i]);
            dfs(nums, list, result);
            list.remove(list.size() - 1);
        }
    }
}

算了好几次超时
import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        result.add(new LinkedList<Integer>());
        for(int i=0; i<nums.length; i++){
            List<List<Integer>>  current = new ArrayList<>();
            for(List<Integer> list:result){
                for(int j=0; j<=list.size(); j++){
                    list.add(j,nums[i]);   //插入法
                    List<Integer> tmp = new ArrayList<>(list);
                    current.add(tmp);
                    list.remove(j);   //节省内存
                }
            }
            result = current;
        }
        return result;
    }

    public static final  void swap(int[] a, int i, int j){
        if(i<0 || j<0 || i>=a.length || j>=a.length)  return;
        int t=a[i];
        a[i]=a[j];
        a[j]=t;
    }
}
递归实现  
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        List<Integer> list = new ArrayList<>();
                if (nums == null || nums.length == 0) return result;
        dfs(nums,list,result);
        return result;
    }

    public static void dfs(int[] nums,List<Integer> list, List<List<Integer>> result ){
        if(list.size()==nums.length){
            result.add(new ArrayList<Integer>(list));   //必须强制声明转换，然而写不报错，实际为空
            return;
        }
        for(int i=0; i<nums.length; i++){
            if(list.contains(nums[i]))  continue;    //穷据递归插入
            list.add(nums[i]);
            dfs(nums,list,result);
            list.remove(list.size()-1);
        }
    }

4. Subsets

Given a set of distinct integers, nums, return all possible subsets (the power set).

Note: The solution set must not contain duplicate subsets.

Example:

Input: nums = [1,2,3]
Output:
[
  [3],
  [1],
  [2],
  [1,2,3],
  [1,3],
  [2,3],
  [1,2],
  []
]
位图表示有没有
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        int n = nums.length;
        int k =(int)Math.pow(2,n);
        List<List<Integer>> sets=new LinkedList<>();
        for(int i=0; i<k; i++){
            List<Integer> list=new LinkedList<>();
            for(int j=0; j<n; j++){
                if( (i&(1<<j))!=0 )  list.add(nums[j]);
            }
            sets.add(list);
        }
        return  sets;
    }
}

5.Word Search

Given a 2D board and a word, find if the word exists in the grid.

The word can be constructed from letters of sequentially adjacent cell, where "adjacent" cells are those horizontally or vertically neighboring. The same letter cell may not be used more than once.

Example:

board =
[
  ['A','B','C','E'],
  ['S','F','C','S'],
  ['A','D','E','E']
]

Given word = "ABCCED", return true.
Given word = "SEE", return true.
Given word = "ABCB", return false.


class Solution {
    public boolean exist(char[][] board, String word) {
        if(board==null)  return false;;
        int m=board.length;
        if(m==0)  return false;
        int n=board[0].length;
        if(n==0)  return false;
        boolean[][] visited=new boolean[m][n];
        char[] chars=word.toCharArray();
        for(int i=0; i<m; i++){
            for(int j=0; j<n; j++){
                if(find(board,visited,i,j,chars,0))  return true;
            }
        }
        return false;
    }
    boolean find(char[][] board, boolean[][] visited,int row,int column,char[] chars, int index){
        if(index==chars.length)  return true;
        int m=board.length;
        int n=board[0].length;
        if(row<0 || column<0 || row>=m ||column>=n || visited[row][column] || board[row][column]!=chars[index] )   return false;
        visited[row][column] =true;
        if(find(board,visited,row-1,column,chars,index+1))  return true;
        if(find(board,visited,row+1,column,chars,index+1))  return true;
        if(find(board,visited,row,column-1,chars,index+1))  return true;
        if(find(board,visited,row,column+1,chars,index+1))  return true;
        visited[row][column]=false;
        return false;
    }
}

File: hard/BackTracking.py
1.Palindrome Partitioning

Given a string s, partition s such that every substring of the partition is a palindrome.

Return all possible palindrome partitioning of s.

Example:

Input: "aab"
Output:
[
  ["aa","b"],
  ["a","a","b"]
]

from typing import List


# !copy in python

class Solution:

    def partition(self, s: str) -> List[List[str]]:
        def isPalinDrome(s, i, j):
            while (i <= j):
                if s[i] != s[j]:
                    return False
                i += 1
                j -= 1
            return True

        n=len(s)
        # flag=[[False]*n]*n
        flag=[]
        for i in range(n):
            row=[False]*n
            for j in range(i,n):
                row[j]=(isPalinDrome(s, i, j))
                # flag[i][j]=isPalinDrome(s, i, j)
                # print(i,j,flag[i][j],flag)
            flag.append(row)
        print(flag)

        def bt(s,i,seq):
            if len(seq)>0 and i==len(s):
                re.append(seq)
                return
            for j in range(i,len(s)):
                if flag[i][j]:
                    bt(s,j+1,seq+[s[i:j+1]])
                # del seq[-1]

        re=[]
        # seq=[]
        bt(s,0,[])
        return re

sln=Solution()
d=sln.partition('aab')
print(d)


2.Word Search II

Given a 2D board and a list of words from the dictionary, find all words in the board.

Each word must be constructed from letters of sequentially adjacent cell, where "adjacent" cells are those horizontally or vertically neighboring. The same letter cell may not be used more than once in a word.

 

Example:

Input: 
board = [
  ['o','a','a','n'],
  ['e','t','a','e'],
  ['i','h','k','r'],
  ['i','f','l','v']
]
words = ["oath","pea","eat","rain"]

Output: ["eat","oath"]

 

Note:

    All inputs are consist of lowercase letters a-z.
    The values of words are distinct.

2. Word Search II

Given a 2D board and a list of words from the dictionary, find all words in the board.

Each word must be constructed from letters of sequentially adjacent cell, where "adjacent" cells are those horizontally or vertically neighboring. The same letter cell may not be used more than once in a word.

 

Example:

Input: 
board = [
  ['o','a','a','n'],
  ['e','t','a','e'],
  ['i','h','k','r'],
  ['i','f','l','v']
]
words = ["oath","pea","eat","rain"]

Output: ["eat","oath"]

 

Note:

    All inputs are consist of lowercase letters a-z.
    The values of words are distinct.

   Hide Hint #1  
You would need to optimize your backtracking to pass the larger test. Could you stop backtracking earlier?
   Hide Hint #2  
If the current candidate does not exist in all words' prefix, you could stop backtracking immediately. What kind of data structure could answer such query efficiently? Does a hash table work? Why or why not? How about a Trie? If you would like to learn how to implement a basic trie, please work on this problem: Implement Trie (Prefix Tree) first.
from typing import List


class Trie:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.children = [None] * 26
        self.tail = False

    def insert(self, word: str) -> None:
        """
        Inserts a word into the trie.
        """
        if len(word) == 0:
            return
        idx = ord(word[0]) - ord('a')
        if not self.children[idx]:
            self.children[idx] = Trie()
        if len(word) == 1:
            self.children[idx].tail = True
        else:
            self.children[idx].insert(word[1:])

    def search(self, word: str) -> bool:
        """
        Returns if the word is in the trie.
        """
        if len(word) == 0:
            return True
        idx = ord(word[0]) - ord('a')
        if not self.children[idx]:
            return False
        if len(word) == 1:
            return self.children[idx].tail  # switch for only once

        return self.children[idx].search(word[1:])

    def startsWith(self, prefix: str) -> bool:
        """
        Returns if there is any word in the trie that starts with the given prefix.
        """
        word = prefix
        if len(word) == 0:
            return True
        idx = ord(word[0]) - ord('a')
        if self.children[idx] is None:
            return False
        return self.children[idx].startsWith(word[1:])


class Solution:
    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        trie = Trie()
        for word in words:
            trie.insert(word)

        def bt(board, m, n, i, j, word, visited):
            if i < 0 or j < 0 or i >= m or j >= n or visited[i][j]:
                return
            word += board[i][j]
            if not trie.startsWith(word):
                return
            if trie.search(word):
                print(word)
                re.append(word)

                # return True

            visited[i][j] = True
            # print(i,j,word,visited)
            bt(board, m, n, i - 1, j, word, visited)
            bt(board, m, n, i + 1, j, word, visited)
            bt(board, m, n, i, j - 1, word, visited)
            bt(board, m, n, i, j + 1, word, visited)
            word=word[0:-1]
            visited[i][j]=False

        m = len(board)
        if m == 0:
            return []
        n = len(board[0])
        if n == 0:
            return []
        if len(words) == 0:
            return words
        re = []

        def gen_mask(m, n):
            mask = [0] * m
            # row=[False]*n
            # board=[row for i in range(m)]
            for i in range(m):
                mask[i] = [False] * n
            return mask

        # mask=gen_mask(m,n)

        for i in range(m):
            for j in range(n):
                if trie.children[ord(board[i][j]) - ord('a')]:
                    bt(board, m, n, i, j, '', gen_mask(m, n))
        return list(set(re))


if __name__ == '__main__':
    board = [
        ['o', 'a', 'a', 'n'],
        ['e', 't', 'a', 'e'],
        ['i', 'h', 'k', 'r'],
        ['i', 'f', 'l', 'v']
    ]
    words = ["oath", "pea", "eat", "rain"]

    # board = [["a","b"],["c","d"]]
    # words = ["cdba"]

    board = [["b", "a", "a", "b", "a", "b"], ["a", "b", "a", "a", "a", "a"], ["a", "b", "a", "a", "a", "b"],
             ["a", "b", "a", "b", "b", "a"], ["a", "a", "b", "b", "a", "b"], ["a", "a", "b", "b", "b", "a"],
             ["a", "a", "b", "a", "a", "b"]]
    words = ["bbaabaabaaaaabaababaaaaababb", "aabbaaabaaabaabaaaaaabbaaaba", "babaababbbbbbbaabaababaabaaa",
             "bbbaaabaabbaaababababbbbbaaa", "babbabbbbaabbabaaaaaabbbaaab", "bbbababbbbbbbababbabbbbbabaa",
             "babababbababaabbbbabbbbabbba", "abbbbbbaabaaabaaababaabbabba", "aabaabababbbbbbababbbababbaa",
             "aabbbbabbaababaaaabababbaaba", "ababaababaaabbabbaabbaabbaba", "abaabbbaaaaababbbaaaaabbbaab",
             "aabbabaabaabbabababaaabbbaab", "baaabaaaabbabaaabaabababaaaa", "aaabbabaaaababbabbaabbaabbaa",
             "aaabaaaaabaabbabaabbbbaabaaa", "abbaabbaaaabbaababababbaabbb", "baabaababbbbaaaabaaabbababbb",
             "aabaababbaababbaaabaabababab", "abbaaabbaabaabaabbbbaabbbbbb", "aaababaabbaaabbbaaabbabbabab",
             "bbababbbabbbbabbbbabbbbbabaa", "abbbaabbbaaababbbababbababba", "bbbbbbbabbbababbabaabababaab",
             "aaaababaabbbbabaaaaabaaaaabb", "bbaaabbbbabbaaabbaabbabbaaba", "aabaabbbbaabaabbabaabababaaa",
             "abbababbbaababaabbababababbb", "aabbbabbaaaababbbbabbababbbb", "babbbaabababbbbbbbbbaabbabaa"]

    # words=["ab","cb","ad","bd","ac","ca","da","bc","db","adcb","dabc","abb","acb"]
    # words=['ab']
    # board=[["a","b"],["c","d"]]
    sln = Solution()
    d = sln.findWords(board, words)
    print(d)

3.Remove Invalid Parentheses

Remove the minimum number of invalid parentheses in order to make the input string valid. Return all possible results.

Note: The input string may contain letters other than the parentheses ( and ).

Example 1:

Input: "()())()"
Output: ["()()()", "(())()"]

Example 2:

Input: "(a)())()"
Output: ["(a)()()", "(a())()"]

Example 3:

Input: ")("
Output: [""]

   Hide Hint #1  
Since we don't know which of the brackets can possibly be removed, we try out all the options!
   Hide Hint #2  
We can use recursion to try out all possibilities for the given expression. For each of the brackets, we have 2 options:

    We keep the bracket and add it to the expression that we are building on the fly during recursion.
    OR, we can discard the bracket and move on. 

   Hide Hint #3  
The one thing all these valid expressions have in common is that they will all be of the same length i.e. as compared to the original expression, all of these expressions will have the same number of characters removed. Can we somehow find the number of misplaced parentheses and use it in our solution?
   Hide Hint #4  
The one thing all these valid expressions have in common is that they will all be of the same length i.e. as compared to the original expression, all of these expressions will have the same number of characters removed. Can we somehow find the number of misplaced parentheses and use it in our solution?
   Hide Hint #5  
For every left parenthesis, we should have a corresponding right parenthesis. We can make use of two counters which keep track of misplaced left and right parenthesis and in one iteration we can find out these two values.

0 1 2 3 4 5 6 7
( ) ) ) ( ( ( )  
i = 0, left = 1, right = 0
i = 1, left = 0, right = 0
i = 2, left = 0, right = 1
i = 3, left = 0, right = 2
i = 4, left = 1, right = 2
i = 5, left = 2, right = 2
i = 6, left = 3, right = 2
i = 7, left = 2, right = 2

We have 2 misplaced left and 2 misplaced right parentheses.

from typing import List


class Solution:
    def removeInvalidParentheses(self, s: str) -> List[str]:
        def isParentheses(line):
            count=0
            for x in line:
                if x=='(':
                    count+=1
                if x==')':
                    count-=1
                if count<0:
                    return False
            return  count==0

        def remove(s,i):
            if i+1==len(s):
                return s[:-1]
            return s[0:i]+s[i+1:]

        print(isParentheses(s))
        if isParentheses(s):
            return [s]
        # left,right = 0,0
        # for x in s:
        #     if x == '(':
        #         left += 1
        #     if x == ')':
        #         right -= 1
        #     # print(x,count)

        # c= '(' if count>0 else ')'

        re=set()
        bucket=set([s])
        for i in range(len(s)+1):
            print(i,bucket)
            tmp=set()
            if len(re)>0:
                return list(re)
            for t in bucket:
                for i in range(len(t)):
                    if t[i] not in ['(',')']:
                        continue
                    u=remove(t,i)
                    # del u[i]
                    if isParentheses(u):
                        re.add(u)
                    else:
                        tmp.add(u)
            bucket=tmp
        return ['']

if __name__ == '__main__':
    # line=  "(a)())()"
    # line= "()())()"
    line=")(f"
    line="n"
    sln = Solution()
    d = sln.removeInvalidParentheses(line)
    print(d)

4.Wildcard Matching

Given an input string (s) and a pattern (p), implement wildcard pattern matching with support for '?' and '*'.

'?' Matches any single character.
'*' Matches any sequence of characters (including the empty sequence).

The matching should cover the entire input string (not partial).

Note:

    s could be empty and contains only lowercase letters a-z.
    p could be empty and contains only lowercase letters a-z, and characters like ? or *.

Example 1:

Input:
s = "aa"
p = "a"
Output: false
Explanation: "a" does not match the entire string "aa".

Example 2:

Input:
s = "aa"
p = "*"
Output: true
Explanation: '*' matches any sequence.

Example 3:

Input:
s = "cb"
p = "?a"
Output: false
Explanation: '?' matches 'c', but the second letter is 'a', which does not match 'b'.

Example 4:

Input:
s = "adceb"
p = "*a*b"
Output: true
Explanation: The first '*' matches the empty sequence, while the second '*' matches the substring "dce".

Example 5:

Input:
s = "acdcb"
p = "a*c?b"
Output: false


just match all
from typing import List


class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        i,j=0,0
        istar,jstar=-1,-1
        while(i<len(s)):
            if j<len(p) and (s[i]==p[j] or p[j]=='?'):  #general
                i+=1
                j+=1
            elif j<len(p) and p[j]=='*':  # init start p->
                istar=i
                jstar=j
                j+=1
            elif istar>=0:  # use * again,
                istar+=1
                i=istar
                j=jstar+1
            else:
                return False
        while(j<len(p) and p[j]=='*'):
            j+=1
        return j==len(p)

if __name__ == '__main__':
    s = "acdcb"
    p = "a*c?b"

    # s = "adceb"
    # p = "*a*b"

    sln = Solution()
    d = sln.isMatch(s,p)
    print(d)
4.Regular Expression Matching

Given an input string (s) and a pattern (p), implement regular expression matching with support for '.' and '*'.

'.' Matches any single character.
'*' Matches zero or more of the preceding element.

The matching should cover the entire input string (not partial).

Note:

    s could be empty and contains only lowercase letters a-z.
    p could be empty and contains only lowercase letters a-z, and characters like . or *.

Example 1:

Input:
s = "aa"
p = "a"
Output: false
Explanation: "a" does not match the entire string "aa".

Example 2:

Input:
s = "aa"
p = "a*"
Output: true
Explanation: '*' means zero or more of the preceding element, 'a'. Therefore, by repeating 'a' once, it becomes "aa".

Example 3:

Input:
s = "ab"
p = ".*"
Output: true
Explanation: ".*" means "zero or more (*) of any character (.)".

Example 4:

Input:
s = "aab"
p = "c*a*b"
Output: true
Explanation: c can be repeated 0 times, a can be repeated 1 time. Therefore, it matches "aab".

Example 5:

Input:
s = "mississippi"
p = "mis*is*p*."
Output: false


from typing import List


class Solution:
    def isMatch(self, s: str, p: str) -> bool:
       #  tokens=[]
       #  for i in range(len(p)):
       #      tokens.append(p[i])
       #      if p[i]=='.':
       #          if i+1==len(p) or p[i+1]!='*':
       #              tokens.append('1')
       #
       #  for i in range(1,len(tokens)):
       #      if tokens[i] in ["*",'1']:
       #          tokens[i]=':'+tokens[i]+','
       #  line=''.join(tokens)
       #  if line[-1]==',':
       #      line=line[:-1]
       #  patten=[  x.split(':') for x in line.split(',')]
       # # only match one char

        if len(p)==0:
            return len(s)==0
        match1= len(s)!=0  and  (s[0]==p[0] or p[0]=='.')
        if len(p)>=2 and p[1]=='*':
            return self.isMatch(s,p[2:]) or  ( match1 and self.isMatch(s[1:],p))

        return match1 and self.isMatch(s[1:],p[1:])


if __name__ == '__main__':
    s = "acdcb"
    p = "a*c?b"

    s = "aab"
    p = "c*a*b"

    # s = "mississippi"
    # p = "mis*is*p*."

    sln = Solution()
    d = sln.isMatch(s,p)
    print(d)

